plugins/experimental/certifier/certifier.cc

```diff
@@ -0,0 +1,739 @@
+};
+template <> struct default_delete<X509_REQ> {
+  void
+  operator()(X509_REQ *n)
+  {
+    X509_REQ_free(n);
+  }
+};
+template <> struct default_delete<EVP_PKEY> {
+  void
+  operator()(EVP_PKEY *n)
+  {
+    EVP_PKEY_free(n);
+  }
+};
+template <> struct default_delete<SSL_CTX> {
+  void
+  operator()(SSL_CTX *n)
+  {
+    SSL_CTX_free(n);
+  }
+};
+} // namespace std
+
+/// Name aliases for unique pts to openSSL objects
+using scoped_X509     = std::unique_ptr<X509>;
+using scoped_X509_REQ = std::unique_ptr<X509_REQ>;
+using scoped_EVP_PKEY = std::unique_ptr<EVP_PKEY>;
+using scoped_SSL_CTX  = std::unique_ptr<SSL_CTX>;
+
+class SslLRUList
+{
+private:
+  struct SslData {
+    std::queue<void *> vconnQ;    ///< Current queue of connections waiting for cert
+    std::unique_ptr<SSL_CTX> ctx; ///< Context generated
+    std::unique_ptr<X509> cert;   ///< Cert generated
+    std::string commonName;       ///< SNI
+    bool scheduled = false;       ///< If a TASK thread has been scheduled to generate cert
+                                  ///< The first thread might fail to do so, this flag will help reschedule
+    bool wontdo = false;          ///< if certs not on disk and dynamic gen is disabled
+    /// Doubly Linked List pointers for LRU
+    SslData *prev = nullptr;
+    SslData *next = nullptr;
+
+    SslData() {}
+    ~SslData() { TSDebug(PLUGIN_NAME, "Deleting ssl data for [%s]", commonName.c_str()); }
+  };
+
+  using scoped_SslData = std::unique_ptr<SslLRUList::SslData>;
+
+  // unordered_map is much faster in terms of insertion/lookup/removal
+  // Althogh it uses more space than map, the time efficiency should be more important
+  std::unordered_map<std::string, scoped_SslData> cnDataMap; ///< Map from CN to sslData
+  TSMutex list_mutex;
+
+  int size = 0;
+  int limit;
+  SslData *head = nullptr;
+  SslData *tail = nullptr;
+
+public:
+  SslLRUList(int in_limit = 4096) : limit(in_limit) { list_mutex = TSMutexCreate(); }
+
+  ~SslLRUList() { TSMutexDestroy(list_mutex); }
+
+  // Returns valid ptr to SSL_CTX if successful lookup
+  //         nullptr if not found and create SslData in the map
+  SSL_CTX *
+  lookup_and_create(const char *servername, void *edata, bool &wontdo)
+  {
+    SslData *ssl_data = nullptr;
+    SSL_CTX *ref_ctx  = nullptr;
+    std::string commonName(servername);
+    TSMutexLock(list_mutex);
+    auto dataItr = cnDataMap.find(commonName);
+    /// If such a context exists in dict
+    if (dataItr != cnDataMap.end()) {
+      /// Reuse context if already built, self queued if not
+      if ((ssl_data = dataItr->second.get())->wontdo) {
+        wontdo = true;
+      } else if (ssl_data->ctx) {
+        ref_ctx = ssl_data->ctx.get();
+      } else {
+        ssl_data->vconnQ.push(edata);
+      }
+    } else {
+      /// Add a new ssl_data to dict if not exist
+      scoped_SslData scoped_ssl_data(new SslData);
+      ssl_data             = scoped_ssl_data.get();
+      ssl_data->commonName = commonName;
```

Comment:
Might be good to `std::move(commonName)` and then use `ssl_data->commonName` later instead, avoiding a string duplication.